notebooks/predicting_queries/entity_intent_suggestion.ipynb

```diff
@@ -0,0 +1,348 @@
+    "#taxons_df.to_csv(PROCESSED_DATA+'/taxons.csv')\n",
+    "taxons_df =pd.read_csv(PROJECT_DIR+'/data/processed/taxons.csv')\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": []
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# inner join taxons to session/query data \n",
+    "# We'll end up with multiple taxons per query \n",
+    "# ie each query has potentially many pages viewed prior to the query being made AND pages can have multiple taxons\n",
+    "\n",
+    "query_taxons = queries_df.merge(taxons_df,on='pagePath')\n",
+    "# Split queries up (they're stupidily aggregated in the sql)\n",
+    "query_taxons['query']=query_taxons['search_terms'].map(lambda x: x.split(','))\n",
+    "query_taxons = query_taxons.explode('query')\n",
+    "# Average number of taxons visited per query (around 7)\n",
+    "np.mean(query_taxons.groupby('session_id').size())"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Instead of bag of words, let's try bag of entities\n",
+    "entities = pd.read_csv(PROJECT_DIR+'/data/processed/content_entities.csv')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Some multiword tokens\n",
+    "\n",
+    "\n",
+    "entities['tokens'] = entities['name'].map(lambda x: tuple(x.split()))\n",
+    "tokenizer = MWETokenizer(list(set(entities['tokens'])))\n",
+    "tokenizer.add_mwe([('log', 'in'), ('sign', 'in'), ('sign', 'up')])\n",
+    "entities['entity'] = entities['tokens'].map(lambda x: tokenizer.tokenize(x))\n",
+    "entities"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "entity_set = set(entities.explode('entity')['entity'])\n",
+    "entity_set"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Text preprocessing\n",
+    "translator = str.maketrans('', '', string.punctuation)\n",
+    "\n",
+    "# Tokenise query, explode, and regroup by taxon \n",
+    "# End up with a list of tokens per taxon\n",
+    "query_taxons['tokens'] = query_taxons['query'].map(lambda x: tokenizer.tokenize(x.translate(translator).split()))\n",
+    "query_taxons['entities'] = query_taxons['tokens'].map(lambda x: [token for token in x if token in entity_set])\n",
+    "query_taxons = query_taxons.explode('entities')\n",
+    "query_taxons = query_taxons[['taxon','entities','tokens']]\n",
+    "query_taxons = query_taxons.dropna()\n",
+    "# Group query tokens by taxon so we have a list of tokens per taxon\n",
+    "query_taxons = query_taxons.groupby('taxon').aggregate(lambda x: list(x)).reset_index()"
+   ]
+  },
```

Comment:
This seems to be a repeat of the same code in `simple_intent_suggestion.ipynb` so it'd be good to have this done once, possibly in a separate script and then have that called in. Saves me reviewing stuff. 😅